// react-toggle-dark-mode/dist/index.js
Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@swc/helpers/_/_extends');
var _object_without_properties_loose = require('@swc/helpers/_/_object_without_properties_loose');
var React = require('react');
var web = require('@react-spring/web');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return n;
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

const defaultProperties = {
    dark: {
        circle: {
            r: 9
        },
        mask: {
            cx: '50%',
            cy: '23%'
        },
        svg: {
            transform: 'rotate(40deg)'
        },
        lines: {
            opacity: 0
        }
    },
    light: {
        circle: {
            r: 5
        },
        mask: {
            cx: '100%',
            cy: '0%'
        },
        svg: {
            transform: 'rotate(90deg)'
        },
        lines: {
            opacity: 1
        }
    },
    springConfig: {
        mass: 4,
        tension: 250,
        friction: 35
    }
};
let REACT_TOGGLE_DARK_MODE_GLOBAL_ID = 0;
const mergeThemeProperties = (theme, customTheme)=>({
        circle: _extends._({}, theme.circle, customTheme == null ? void 0 : customTheme.circle),
        mask: _extends._({}, theme.mask, customTheme == null ? void 0 : customTheme.mask),
        svg: _extends._({}, theme.svg, customTheme == null ? void 0 : customTheme.svg),
        lines: _extends._({}, theme.lines, customTheme == null ? void 0 : customTheme.lines)
    });
const resolveAnimationProperties = (animationProperties)=>{
    if (!animationProperties || animationProperties === defaultProperties) {
        return defaultProperties;
    }
    return {
        dark: mergeThemeProperties(defaultProperties.dark, animationProperties.dark),
        light: mergeThemeProperties(defaultProperties.light, animationProperties.light),
        springConfig: _extends._({}, defaultProperties.springConfig, animationProperties.springConfig)
    };
};
const DarkModeSwitch = (_0)=>{
    let { onChange, checked = false, size = 24, animationProperties, moonColor = 'white', sunColor = 'black', style, onClick: onClickProp, role, tabIndex, 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy } = _0, rest = _object_without_properties_loose._(_0, [
        "onChange",
        "checked",
        "size",
        "animationProperties",
        "moonColor",
        "sunColor",
        "style",
        "onClick",
        "role",
        "tabIndex",
        'aria-label',
        'aria-labelledby'
    ]);
    const [id] = React__namespace.useState(()=>{
        REACT_TOGGLE_DARK_MODE_GLOBAL_ID += 1;
        return REACT_TOGGLE_DARK_MODE_GLOBAL_ID;
    });
    const properties = React__namespace.useMemo(()=>resolveAnimationProperties(animationProperties), [
        animationProperties
    ]);
    const { circle, svg, lines, mask } = properties[checked ? 'dark' : 'light'];
    const { springConfig } = properties;
    const svgContainerProps = web.useSpring(_extends._({}, svg, {
        config: springConfig
    }));
    const centerCircleProps = web.useSpring(_extends._({}, circle, {
        config: springConfig
    }));
    const maskedCircleProps = web.useSpring(_extends._({}, mask, {
        config: springConfig
    }));
    const linesProps = web.useSpring(_extends._({}, lines, {
        config: springConfig
    }));
    const toggle = React__namespace.useCallback(()=>onChange(!checked), [
        checked,
        onChange
    ]);
    const onClick = React__namespace.useCallback((event)=>{
        event.stopPropagation();
        onClickProp == null ? void 0 : onClickProp(event);
        if (!event.defaultPrevented) {
            toggle();
        }
    }, [
        onClickProp,
        toggle
    ]);
    const uniqueMaskId = `circle-mask-${id}`;
    return /*#__PURE__*/ React__namespace.createElement("button", _extends._({
        type: "button",
        role: role != null ? role : 'switch',
        "aria-checked": checked,
        "aria-label": ariaLabelledBy ? undefined : ariaLabel != null ? ariaLabel : 'Toggle dark mode',
        "aria-labelledby": ariaLabelledBy,
        tabIndex: tabIndex != null ? tabIndex : 0,
        onClick: onClick,
        style: _extends._({
            cursor: 'pointer',
            padding: 0,
            border: 0,
            background: 'transparent',
            lineHeight: 0
        }, style)
    }, rest), /*#__PURE__*/ React__namespace.createElement(web.animated.svg, {
        xmlns: "http://www.w3.org/2000/svg",
        width: size,
        height: size,
        viewBox: "0 0 24 24",
        color: checked ? moonColor : sunColor,
        fill: "none",
        strokeWidth: "2",
        strokeLinecap: "round",
        strokeLinejoin: "round",
        stroke: "currentColor",
        style: svgContainerProps
    }, /*#__PURE__*/ React__namespace.createElement("mask", {
        id: uniqueMaskId
    }, /*#__PURE__*/ React__namespace.createElement("rect", {
        x: "0",
        y: "0",
        width: "100%",
        height: "100%",
        fill: "white"
    }), /*#__PURE__*/ React__namespace.createElement(web.animated.circle, {
        // @ts-ignore
        style: maskedCircleProps,
        r: "9",
        fill: "black"
    })), /*#__PURE__*/ React__namespace.createElement(web.animated.circle, {
        cx: "12",
        cy: "12",
        fill: checked ? moonColor : sunColor,
        style: centerCircleProps,
        mask: `url(#${uniqueMaskId})`
    }), /*#__PURE__*/ React__namespace.createElement(web.animated.g, {
        stroke: "currentColor",
        style: linesProps
    }, /*#__PURE__*/ React__namespace.createElement("line", {
        x1: "12",
        y1: "1",
        x2: "12",
        y2: "3"
    }), /*#__PURE__*/ React__namespace.createElement("line", {
        x1: "12",
        y1: "21",
        x2: "12",
        y2: "23"
    }), /*#__PURE__*/ React__namespace.createElement("line", {
        x1: "4.22",
        y1: "4.22",
        x2: "5.64",
        y2: "5.64"
    }), /*#__PURE__*/ React__namespace.createElement("line", {
        x1: "18.36",
        y1: "18.36",
        x2: "19.78",
        y2: "19.78"
    }), /*#__PURE__*/ React__namespace.createElement("line", {
        x1: "1",
        y1: "12",
        x2: "3",
        y2: "12"
    }), /*#__PURE__*/ React__namespace.createElement("line", {
        x1: "21",
        y1: "12",
        x2: "23",
        y2: "12"
    }), /*#__PURE__*/ React__namespace.createElement("line", {
        x1: "4.22",
        y1: "19.78",
        x2: "5.64",
        y2: "18.36"
    }), /*#__PURE__*/ React__namespace.createElement("line", {
        x1: "18.36",
        y1: "5.64",
        x2: "19.78",
        y2: "4.22"
    }))));
};

exports.DarkModeSwitch = DarkModeSwitch;
exports.defaultProperties = defaultProperties;
