// react-toggle-dark-mode/dist/react-toggle-dark-mode.esm.js
import { _ as _$1 } from '@swc/helpers/_/_extends';
import { _ } from '@swc/helpers/_/_object_without_properties_loose';
import * as React from 'react';
import { useSpring, animated } from '@react-spring/web';

const defaultProperties = {
    dark: {
        circle: {
            r: 9
        },
        mask: {
            cx: '50%',
            cy: '23%'
        },
        svg: {
            transform: 'rotate(40deg)'
        },
        lines: {
            opacity: 0
        }
    },
    light: {
        circle: {
            r: 5
        },
        mask: {
            cx: '100%',
            cy: '0%'
        },
        svg: {
            transform: 'rotate(90deg)'
        },
        lines: {
            opacity: 1
        }
    },
    springConfig: {
        mass: 4,
        tension: 250,
        friction: 35
    }
};
let REACT_TOGGLE_DARK_MODE_GLOBAL_ID = 0;
const mergeThemeProperties = (theme, customTheme)=>({
        circle: _$1({}, theme.circle, customTheme == null ? void 0 : customTheme.circle),
        mask: _$1({}, theme.mask, customTheme == null ? void 0 : customTheme.mask),
        svg: _$1({}, theme.svg, customTheme == null ? void 0 : customTheme.svg),
        lines: _$1({}, theme.lines, customTheme == null ? void 0 : customTheme.lines)
    });
const resolveAnimationProperties = (animationProperties)=>{
    if (!animationProperties || animationProperties === defaultProperties) {
        return defaultProperties;
    }
    return {
        dark: mergeThemeProperties(defaultProperties.dark, animationProperties.dark),
        light: mergeThemeProperties(defaultProperties.light, animationProperties.light),
        springConfig: _$1({}, defaultProperties.springConfig, animationProperties.springConfig)
    };
};
const DarkModeSwitch = (_0)=>{
    let { onChange, checked = false, size = 24, animationProperties, moonColor = 'white', sunColor = 'black', style, onClick: onClickProp, role, tabIndex, 'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledBy } = _0, rest = _(_0, [
        "onChange",
        "checked",
        "size",
        "animationProperties",
        "moonColor",
        "sunColor",
        "style",
        "onClick",
        "role",
        "tabIndex",
        'aria-label',
        'aria-labelledby'
    ]);
    const [id] = React.useState(()=>{
        REACT_TOGGLE_DARK_MODE_GLOBAL_ID += 1;
        return REACT_TOGGLE_DARK_MODE_GLOBAL_ID;
    });
    const properties = React.useMemo(()=>resolveAnimationProperties(animationProperties), [
        animationProperties
    ]);
    const { circle, svg, lines, mask } = properties[checked ? 'dark' : 'light'];
    const { springConfig } = properties;
    const svgContainerProps = useSpring(_$1({}, svg, {
        config: springConfig
    }));
    const centerCircleProps = useSpring(_$1({}, circle, {
        config: springConfig
    }));
    const maskedCircleProps = useSpring(_$1({}, mask, {
        config: springConfig
    }));
    const linesProps = useSpring(_$1({}, lines, {
        config: springConfig
    }));
    const toggle = React.useCallback(()=>onChange(!checked), [
        checked,
        onChange
    ]);
    const onClick = React.useCallback((event)=>{
        event.stopPropagation();
        onClickProp == null ? void 0 : onClickProp(event);
        if (!event.defaultPrevented) {
            toggle();
        }
    }, [
        onClickProp,
        toggle
    ]);
    const uniqueMaskId = `circle-mask-${id}`;
    return /*#__PURE__*/ React.createElement("button", _$1({
        type: "button",
        role: role != null ? role : 'switch',
        "aria-checked": checked,
        "aria-label": ariaLabelledBy ? undefined : ariaLabel != null ? ariaLabel : 'Toggle dark mode',
        "aria-labelledby": ariaLabelledBy,
        tabIndex: tabIndex != null ? tabIndex : 0,
        onClick: onClick,
        style: _$1({
            cursor: 'pointer',
            padding: 0,
            border: 0,
            background: 'transparent',
            lineHeight: 0
        }, style)
    }, rest), /*#__PURE__*/ React.createElement(animated.svg, {
        xmlns: "http://www.w3.org/2000/svg",
        width: size,
        height: size,
        viewBox: "0 0 24 24",
        color: checked ? moonColor : sunColor,
        fill: "none",
        strokeWidth: "2",
        strokeLinecap: "round",
        strokeLinejoin: "round",
        stroke: "currentColor",
        style: svgContainerProps
    }, /*#__PURE__*/ React.createElement("mask", {
        id: uniqueMaskId
    }, /*#__PURE__*/ React.createElement("rect", {
        x: "0",
        y: "0",
        width: "100%",
        height: "100%",
        fill: "white"
    }), /*#__PURE__*/ React.createElement(animated.circle, {
        // @ts-ignore
        style: maskedCircleProps,
        r: "9",
        fill: "black"
    })), /*#__PURE__*/ React.createElement(animated.circle, {
        cx: "12",
        cy: "12",
        fill: checked ? moonColor : sunColor,
        style: centerCircleProps,
        mask: `url(#${uniqueMaskId})`
    }), /*#__PURE__*/ React.createElement(animated.g, {
        stroke: "currentColor",
        style: linesProps
    }, /*#__PURE__*/ React.createElement("line", {
        x1: "12",
        y1: "1",
        x2: "12",
        y2: "3"
    }), /*#__PURE__*/ React.createElement("line", {
        x1: "12",
        y1: "21",
        x2: "12",
        y2: "23"
    }), /*#__PURE__*/ React.createElement("line", {
        x1: "4.22",
        y1: "4.22",
        x2: "5.64",
        y2: "5.64"
    }), /*#__PURE__*/ React.createElement("line", {
        x1: "18.36",
        y1: "18.36",
        x2: "19.78",
        y2: "19.78"
    }), /*#__PURE__*/ React.createElement("line", {
        x1: "1",
        y1: "12",
        x2: "3",
        y2: "12"
    }), /*#__PURE__*/ React.createElement("line", {
        x1: "21",
        y1: "12",
        x2: "23",
        y2: "12"
    }), /*#__PURE__*/ React.createElement("line", {
        x1: "4.22",
        y1: "19.78",
        x2: "5.64",
        y2: "18.36"
    }), /*#__PURE__*/ React.createElement("line", {
        x1: "18.36",
        y1: "5.64",
        x2: "19.78",
        y2: "4.22"
    }))));
};

export { DarkModeSwitch, defaultProperties };
